Migrate from rustc-serialize to serde 1.0

ordian opened the pull request:
This is a breaking change.
* decode() is renamed to deserialize()
  for consistency with serde
* docopt_macros is updated to latest nightly

Fixes #128, #213

ordian commented:
Tests on Travis are failing on rust 1.12. 
Any plans to drop support for this old version?

BurntSushi commented on .travis.yml:
```diff
@@ -1,6 +1,5 @@
 language: rust
 rust:
-  - 1.12.0
```
I understand why we can no longer use Rust 1.12 (`serde_derive` requires Rust 1.15 IIRC), but there's a reason why a specific version is tested in Travis: to ensure we don't *accidentally* bump the minimum Rust version required to build docopt. For this reason, please put the minimum version back. I think Rust 1.15 would be suitable.

BurntSushi commented on Cargo.toml:
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "docopt"
-version = "0.7.0"  #:version
+version = "0.8.0"  #:version
```
Please don't bump the version. I'll do that in a separate commit. :-)

BurntSushi commented on docopt_macros/Cargo.toml:
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "docopt_macros"
-version = "0.7.0"  #:version
+version = "0.8.0"  #:version
```
Same as `docopt`'s Cargo.toml. I'll bump this version in a separate commit.

ordian commented:
Fixed.
Bumped minimum rust version to 1.15, although I don't know why anyone would use stable, but not latest.
Also, replaced try! with ? (IIRC stable since 1.13).
Thanks!

BurntSushi commented:
Thanks!

>  although I don't know why anyone would use stable, but not latest.

Because not everyone can upgrade everything all the time. Think about Linux distros, company policies, etc etc.

ordian commented:
Should I revert version bump in readme as well?

BurntSushi commented:
@ordian Yes please, good catch.

ordian commented:
How about enum with struct variant support? E.g.
```rust
#[derive(Debug, Deserialize)]
enum CargoCommand {
    Build {
        flag_debug: bool,
        flag_examples: bool,  
    },
    Run {
        ...
    },
    Test,
    Bench,
    ...
}
```

BurntSushi commented:
@ordian I'd prefer that be separate. It has been discussed quite a bit on the issue tracker, but all the details have fallen out of my head.

BurntSushi commented:
Thanks! This is awesome work. :-)

nixpulvis commented:
:DDDD
